enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,553 @@
+
+Q: What happens if the kcm leader has no access to ccm? Can it continue its work? Will it give up leadership?
+A: KCM and CCM perform independently after the initial migration process. KCM will be unaffected by CCM being down after the initial migration. The CCM cluster operator will be responsible for transitioning to Unavailable or Degraded in a similar situation.
+
+Q: What are the thoughts about certificate management?
+A: At present, only the GCP provider has any CSR approval mechanism built in. This functionality is used to approve `NodeClient` and `NodeServer` certificates for Nodes on GCP.
+This functionality should not be required as OpenShift handles certificate approval via the `cluster-machine-approver` component.
+
+Q: Does every node need a CCM?
+A: No. The cluster only needs one active `CCM` at any time. A `Deployment` will manage the `CCM` pod and will have 2 replicas which will use leader election to nominate an active leader. [Source](https://kubernetes.io/docs/concepts/overview/components/#cloud-controller-manager)
+
+Q: How metrics are affected by the CCM migration?
+A: Currently the `CCM` implementation is using the same set of metrics from the core repository, all existing metrics will still be available post migration. The `CCMO` will be required to create a `ServiceMonitor` to instruct prometheus to collect metrics from the `CCM` pod.
+
+## Other notes
+
+### Removing a deprecated feature
+
+Upstream is currently planning to remove support for the in-tree providers with kubernetes 1.23. The enhancement [describes](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md) steps towards this process. Our goal is to have a working alternative for the in-tree providers ready to be used on-par with the old implementation before upstream remove this feature from core Kubernetes repository.
+
+The `cloud-provider` and `cloud-config` flags from `kubelet` will be removed in 1.23 - [link](https://github.com/kubernetes/kubernetes/pull/90408).
+
+## Timeline
+
+### Upstream
+
+Follow the Kubernetes community discussions and implementation/releases
+
+- vSphere support has graduated to beta in [v1.18](https://github.com/kubernetes/enhancements/issues/670) (Openshift 4.5)
+
+- Azure support goes into beta in [v1.20](https://github.com/kubernetes/enhancements/issues/667) (Openshift 4.7)
+
+### OpenShift
+
+#### OCP 4.8
+
+- AWS on out-of-tree `CCM` tech-preview. `CSI` is optional (via OLM) since [4.6](https://github.com/openshift/enhancements/blob/master/enhancements/storage/csi-driver-install.md#ocp-45-kubernetes-118).
+- OpenStack platform on out-of-tree `CCM` graduates as tech-preview.
+
+#### OCP 4.9
+
+- OpenStack upgrade and installation with the out-of-tree CCM by default in 4.9. (GA)
+
+#### OCP 4.10
+
+- vSphere, `CSI` release plan for tech-preview is 4.7+.
+- Azure.
+- GCP.
```

Comment:
Regarding the timeline - CCM GA must go in sync with CSI migration GA, as CCM implies CSI migration and
we can't have in-tree volume plugin (GA) replaced by CSI driver which is beta / tech preview.

From upstream sig-storage plan, which changes every 3 weeks:

* AWS, GCE, Azure: GA in 1.23 (4.10)
* OpenStack: GA in 1.22 or 1.23 (4.9/4.10)
* vSphere: GA in 1.24 (4.11)

The migration is reaching beta in various releases and from storage team point of view we can do tech preview in 4.8 with GCE, OpenStack and AWS (all of them or a subset, whatever suits you).

Reply:
Update after Kubernetes 1.21 code freeze:

* AWS, GCE, OpenStack, Azure: GA in 1.23 (4.10)
* vSphere: GA in 1.24 (4.11)

From CSI point of view, in 4.8 we need tech preview of OpenStack CSI + CCM migration, AWS is highly recommended (to get feedback of most customers). In 4.9, we may need tech preview of everything except vSphere, to catch upstream GA in 4.10.